lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,967 @@
+}
+
+void TypeLoweringVisitor::visitStmt(WhenOp op) {
+  // The WhenOp itself does not require any lowering, the only value it uses
+  // is a one-bit predicate.  Recursively visit all regions so internal
+  // operations are lowered.
+
+  // Visit operations in the then block.
+  lowerBlock(&op.getThenBlock());
+
+  // If there is no else block, return.
+  if (!op.hasElseRegion())
+    return;
+
+  // Visit operations in the else block.
+  lowerBlock(&op.getElseBlock());
+}
+
+/// Lower memory operations. A new memory is created for every leaf
+/// element in a memory's data type.
+void TypeLoweringVisitor::visitDecl(MemOp op) {
+  // Attempt to get the bundle types.
+  SmallVector<FlatBundleFieldEntry> fields = peelType(op.getDataType());
+  if (fields.empty())
+    return;
+
+  SmallVector<MemOp> newMemories;
+  SmallVector<Value> wireToOldResult;
+  SmallVector<WireOp> oldPorts;
+
+  // Wires for old ports
+  for (unsigned int index = 0, end = op.getNumResults(); index < end; ++index) {
+    auto result = op.getResult(index);
+    auto wire = builder->create<WireOp>(
+        result.getType(),
+        (op.name() + "_" + op.getPortName(index).getValue()).str());
+    oldPorts.push_back(wire);
+    result.replaceAllUsesWith(wire.getResult());
+  }
+
+  // Memory for each field
+  for (auto field : fields)
+    newMemories.push_back(
+        cloneMemWithNewType(builder, op, field.type, field.suffix));
+
+  // Hook up the new memories to the wires the old memory was replaced with.
+  for (size_t index = 0, rend = op.getNumResults(); index < rend; ++index) {
+    auto result = oldPorts[index];
+    auto rType = result.getType().cast<BundleType>();
+    for (size_t fieldIndex = 0, fend = rType.getNumElements();
+         fieldIndex != fend; ++fieldIndex) {
+      auto name = rType.getElement(fieldIndex).name.getValue();
+      auto oldField = builder->create<SubfieldOp>(result, name);
+      // data and mask depend on the memory type which was split.  They can also
+      // go both directions, depending on the port direction.
+      if (name == "data" || name == "mask") {
+        for (auto field : fields) {
+          auto realOldField = getSubWhatever(oldField, field.index);
+          auto newField = getSubWhatever(
+              newMemories[field.index].getResult(index), fieldIndex);
+          if (rType.getElement(fieldIndex).isFlip)
+            std::swap(realOldField, newField);
+          builder->create<ConnectOp>(newField, realOldField);
+        }
+      } else {
+        for (auto mem : newMemories) {
+          auto newField =
+              builder->create<SubfieldOp>(mem.getResult(index), name);
+          builder->create<ConnectOp>(newField, oldField);
+        }
+      }
+    }
+  }
+  opsToRemove.push_back(op);
+}
+
+void TypeLoweringVisitor::visitDecl(FExtModuleOp extModule) {
+  ImplicitLocOpBuilder theBuilder(extModule.getLoc(), context);
+  builder = &theBuilder;
+  
+  // Top level builder
+  OpBuilder builder(context);
```

Comment:
Are the first two lines unnecessary?

Reply:
no, one builder is top-level, one builder is at the operation.